bpfman-operator/controllers/bpfman-agent/common.go

```diff
@@ -447,8 +436,75 @@ func reconcileProgram(ctx context.Context,
 		}
 	}
 
-	// We didn't already return something else, so there's nothing to do
-	r.Logger.Info("Nothing to do for this program")
+	// We're done reconciling the expected programs.  If any unexpected programs
+	// exist, delete them and return the result.
+	return handleProgDelete(ctx, r, common, rec, existingPrograms, programMap, isNodeSelected, isBeingDeleted, mapOwnerStatus)
+}
+
+// reconcileProgram is called by ALL *Program controllers, and contains much of
+// the core logic for taking *Program objects, turning them into bpfProgram
+// object(s), and ultimately telling the custom controller types to load real
+// bpf programs on the node via bpfman. Additionally it acts as a central point for
+// interacting with the K8s API. This function will exit if any action is taken
+// against the K8s API. If the function returns a retry boolean and error, the
+// reconcile will be retried based on a default 5 second interval if the retry
+// boolean is set to `true`.
```

Comment:
@astoycos, you made some good comments on the function and I it to try to address them.  As part of this, I broke out the handling of the case were we are deleting programs and the case where we are not.  I also reused the delete function to delete unexpected programs.  I'm sure it can be broken into more functional components, but this is a start.

Reply:
+1 One nit is that `handleProgramCreate` is also handling updates right? Maybe change it to just `handleProgram`?

Reply:
Naming is hard.  I thought about this a bit when I was doing it, but couldn't left it as handleProgramCreate.  It's really handleProgCreateAndUpdate, or handleProgNotBeingDeleted.  I'll think about it.